### PR TITLE
connectors-ci: add `finalize_build` logic to handle custom Dockerfiles

### LINKED ISSUE #26426: connector-ci: implement custom build logic for some java connectors
Connectors like `destination-s3` [have custom Dockerfile for installation of system dependencies](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/destination-s3/Dockerfile#L1).
This is problematic because our Dagger pipeline builds all the java connectors the same way with the logic declared [here](https://github.com/airbytehq/airbyte/blob/master/tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py#L649).

As the number of connector not following the default Dockerfile for java connectors I suggest to implement custom special build case for the specific connectors.

In the same work we shall add a comment to all the java connectors Dockerfile that these file are not used to build the connectors we publish.

### PR DESCRIPTION
**Pull request by alafanechere**
## What
Closes #26426
Java connectors are not built with Dockerfiles anymore now that we switched to Dagger.
But some connectors have a custom Dockerfile that install system dependencies are add custom env variable.
This PR is a suggested approach to handle these exceptional cases.

## How
* A `finalize_build.sh` or a `finalize_build.py` file can be created in the connector folder. Developers are free to pick the format they prefer. The `finalize_build` python function allows direct interaction with the Dagger container being built, which can be convenient for docker specific instructions and improved caching. 
* If this file is present the `with_airbyte_python_connector` or `with_airbyte_java_connector` function will run the bash script or the `finalize_build` Python function on the container that was built with the common procedure.

I added a deprecation warning in order to make our developers aware that Dockerfile changes currently have no effect on images we currently publish with the automated pipelines. 
We can't delete the Dockerfiles at the moment because they're still use for `/test` - but once the tests powered by dagger are live we can proceed to their deletion.

For a "real-life" example I've added:
- a `finalize_build.sh` script to destination-s3 to install LZO compression.
- a `finalize_build.py` script to destination-iceberg to set the custom JAVA_OPTS env var this connector requires

⚠️ Please note this PR does not fix destination-s3 or destination-iceberg:
* The `finalize_build.sh` script for `destination-s3` exits with 1 status code on ARM64 architecutre
* The `destination-iceberg` connector build has been broken for a long time (probably since monorepo split) and relies on Java library that are not on the airbyte repo anymore... 

## Recommended reading order:
* https://github.com/airbytehq/airbyte/blob/d17ec254843a423076e19914565db5f65a7e0679/tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py#L686
* https://github.com/airbytehq/airbyte/blob/d17ec254843a423076e19914565db5f65a7e0679/airbyte-integrations/connectors/destination-s3/finalize_build.sh#L1
* https://github.com/airbytehq/airbyte/blob/d17ec254843a423076e19914565db5f65a7e0679/airbyte-integrations/connectors/destination-iceberg/finalize_build.py#L6
* https://github.com/airbytehq/airbyte/pull/26489/commits/d17ec254843a423076e19914565db5f65a7e0679
## 🚨 User Impact 🚨
No impact as the build logic for connector with "normal" Dockerfiles was not altered. I'll publish a java and a python connector pre-release for sanity.
Update: [confirmed](https://github.com/airbytehq/airbyte/actions/runs/5069795932) that the "normal" connector publish is working. The `destination-s3` failure is related to an error in the `finalize_build.sh` script I pinged @edgao about [here](https://github.com/airbytehq/airbyte/pull/26489#discussion_r1204221382).

### CODE REVIEW

File: airbyte-integrations/connectors/destination-s3/finalize_build.sh [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#!/bin/bash
+set -e
+
+ARCH=$(uname -m)
+
+if [ "$ARCH" == "x86_64" ] || [ "$ARCH" = "amd64" ]; then
+  echo "$ARCH"
+  yum install lzop lzo lzo-dev -y
+
+# alanechere: I'm not sure we need this custom install of lzo anymore. Using the yum install above works in the build context.
```
Comment (alafanechere) [inline airbyte-integrations/connectors/destination-s3/finalize_build.sh:10]: @edgao I was not able to run this script successfully. The `sdk install java 8.0.342-librca` command exits with a 1 status code. Considering the current code freeze on this connector I think it's fine to merge this PR with a "broken" script. I'll let your team figure the changes required to support LZO on all architectures. I'm even not sure the custom install for ARM64 is required. I tried a simple  `yum install lzop lzo lzo-dev -y` on the ARM64 variant and it worked.

You can reproduce my failure on this branch by running:
`airbyte-ci connectors --name=destination-s3 build`

I'll also let you decide if you prefer to use a bash script or the `finalize_build.py` module with a `finalize_build` (please check the `destination-iceberg/finalize_build.py` for an example). The 🐍 way is better for improved caching and access to the Dagger API. But I also understand that running this script as it is is convenient.

**Review by evantahler — APPROVED** [review]
Nice, and the comments in the Java dockerfiles are great too

File: airbyte-integrations/connectors/destination-iceberg/finalize_build.py
Comment (bnchrch) [inline airbyte-integrations/connectors/destination-iceberg/finalize_build.py:1]: super small nit:

I'm accustomed to seeing `post_`, `pre_`, or `custom` or `override` in cases like these
1. post_install or pre_install with [npm](https://docs.npmjs.com/cli/v9/using-npm/scripts)
2. post_deploy or pre_deploy on [heroku](https://devcenter.heroku.com/articles/release-phase#review-apps-and-the-postdeploy-script)

Would some thing like `post_build`, `custom_build`, `build_script` fit better?

File: tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py [hunk 2]
```diff
@@ -667,20 +668,63 @@ async def with_airbyte_java_connector(context: ConnectorContext, connector_java_
...
+            connector_container = await connector_finalize_module.finalize_build(context, connector_container)
+        except AttributeError:
+            raise Exception("Connector has a finalize_build.py script but it doesn't have a finalize_build function.")
+
+    if has_finalize_bash_script:
+        context.logger.info(f"{context.connector.technical_name} has finalize_build.sh script, running it to finalize build...")
+        connector_container = (
+            connector_container.with_file("/tmp/finalize_build.sh", connector_dir_with_finalize_script.file("finalize_build.sh"))
+            .with_entrypoint("sh")
+            .with_exec(["/tmp/finalize_build.sh"])
```
Comment (bnchrch) [inline tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py:719]:
```suggestion
            .with_exec(["sh", "/tmp/finalize_build.sh"])
```

File: tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py [hunk 3]
```diff
@@ -667,20 +668,63 @@ async def with_airbyte_java_connector(context: ConnectorContext, connector_java_
...
         context.dagger_client.container(platform=build_platform)
         .with_mounted_cache("/root/.cache/pip", pip_cache)
         .build(context.get_connector_dir())
         .with_label("io.airbyte.version", context.metadata["dockerImageTag"])
         .with_label("io.airbyte.name", context.metadata["dockerRepository"])
     )
+    return await finalize_build(context, connector_container)
+
+
+async def finalize_build(context: ConnectorContext, connector_container: Container) -> Container:
```
Comment (bnchrch) [inline tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py:686]: nit: Lets put this in the docs!

**Review by bnchrch — APPROVED** [review]
Well done!

Just a couple very small suggestions.

Loving how extensible this makes our pipelines

File: airbyte-integrations/connectors/destination-s3/finalize_build.sh (hunk 1 above)
Comment (edgao) [reply airbyte-integrations/connectors/destination-s3/finalize_build.sh:10]: amazing. yeah I'll just submit an issue to figure this out, not a blocker for this pr.

File: tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py [hunk 4]
```diff
@@ -667,20 +668,63 @@ async def with_airbyte_java_connector(context: ConnectorContext, connector_java_
...
+    """Finalize build by running finalize_build.sh or finalize_build.py if present in the connector directory."""
+    connector_dir_with_finalize_script = context.get_connector_dir(include=["finalize_build.sh", "finalize_build.py"])
+    finalize_scripts = await connector_dir_with_finalize_script.entries()
+    if not finalize_scripts:
+        return connector_container
+
+    # We don't want finalize scripts to override the entrypoint so we keep it in memory to reset it after finalization
```
Comment (edgao) [inline tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py:693]: just curious, why don't we want the scripts to override the entrypoint?

**Review by edgao — APPROVED** [review]
lgtm, submitted https://github.com/airbytehq/airbyte/issues/26521 to get dest-s3 into good state.

File: tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py (hunk 4 above)
Comment (alafanechere) [reply tools/ci_connector_ops/ci_connector_ops/pipelines/actions/environments.py:693]: I might reconsider it for specific upcoming cases, but I've not seen connector specifically overriding the classic airbyte entrypoint (except your "normalization in container" usecase, which is handled directly in the main code path you contributed to)

**Pull request merged**